File: 147-insertion-sort-list/ken/147.py
# Definition for singly-linked list.
# class ListNode(object):
#	 def __init__(self, x):
#		 self.val = x
#		 self.next = None

class Solution(object):
	def insertionSortList(self, head):
		"""
		:type head: ListNode
		:rtype: ListNode
		"""

		if head is None:
			return None

		new_head = head
		cur = head.next
		new_head.next = None

		while cur:
			next = cur.next

			if cur.val <= new_head.val:
				cur.next = new_head
				new_head = cur
			else:
				i = new_head
				while i:
					if i.next is None or cur.val <= i.next.val:
						cur.next = i.next
						i.next = cur
						break
					i = i.next

			cur = next

		return new_head

File: 147-insertion-sort-list/leochang/147.cpp
#include <cstdio>
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* insertionSortList(ListNode* head) {
		if (!head) {
			return head;
		}

		ListNode* current_head = head;
		ListNode* next_node = NULL;
		ListNode* new_head = new ListNode(0);

		while (current_head) {
			next_node = current_head->next;
			ListNode* prev_node = new_head;

			while (prev_node->next && prev_node->next->val < current_head->val) {
				prev_node = prev_node->next;
			}

			current_head->next = prev_node->next;
			prev_node->next = current_head;
			current_head = next_node;
		}

		return new_head->next;
	}
};
